packages/walletconnect-connector/src/WalletConnectConnector.ts

```diff
@@ -197,49 +221,128 @@ export class WalletConnectConnector extends PredicateConnector {
-    return new Promise((resolve) => {
+    const result = await new Promise<boolean>((resolve, reject) => {
       this.web3Modal.open();
       const wagmiConfig = this.getWagmiConfig();
       const unsub = this.web3Modal.subscribeEvents(async (event) => {
+        const requestValidations = () => {
+          this.requestValidations()
+            .then(() => resolve(true))
+            .catch((err) => reject(err))
+            .finally(() => unsub());
+        };
+
         switch (event.data.event) {
           case 'MODAL_OPEN':
             if (wagmiConfig) {
               const account = getAccount(wagmiConfig);
               if (account?.isConnected) {
-                this.web3Modal.close();
-                resolve(true);
                 unsub();
+                this.web3Modal.close();
+                requestValidations();
                 break;
               }
             }
             // Ensures that the WC Web3Modal config is applied over pre-existing states (e.g. Solan Connect Web3Modal)
             this.createModal();
             break;
           case 'CONNECT_SUCCESS': {
-            resolve(true);
-            unsub();
+            requestValidations();
             break;
           }
           case 'MODAL_CLOSE':
           case 'CONNECT_ERROR': {
+            if (wagmiConfig) {
+              const account = getAccount(wagmiConfig);
+              if (account) {
+                requestValidations();
+                break;
+              }
+            }
             resolve(false);
             unsub();
             break;
           }
         }
       });
     });
+    return result;
+  }
+
+  private async getAccountValidations(
+    accounts: `0x${string}`[] | string[],
+  ): Promise<boolean[]> {
+    return Promise.all(
+      accounts.map(async (a) => {
+        const isValidated = await this.storage.getItem(
+          `SIGNATURE_VALIDATION_${a}`,
+        );
+        return isValidated === 'true';
+      }),
+    );
+  }
+
+  private async accountHasValidation(
+    account: `0x${string}` | string | undefined,
+  ) {
+    if (!account) return false;
+    const [hasValidate] = await this.getAccountValidations([account]);
+    return hasValidate;
+  }
+
+  async requestValidations() {
+    const wagmiConfig = this.getWagmiConfig();
+    if (!wagmiConfig) {
+      throw new Error('Wagmi config not found');
+    }
+    const account = getAccount(wagmiConfig);
+    const { addresses } = account;
+    for (const address of addresses || []) {
+      await this.requestValidation(address)
+        .then(() => {
+          this.handleConnect(account);
+        })
+        .catch((err) => {
+          this.disconnect();
+          throw err;
+        });
+    }
+  }
+
+  async requestValidation(address?: string) {
+    return new Promise(async (resolve, reject) => {
+      // Disconnect if user dosen't provide signature in time
+      const validationTimeout = setTimeout(() => {
+        reject(
```

Comment:
If the connection has to wait for validation there's no need for a timeout

Reply:
This is a timeout to avoid been pending forever if some error occur. In the case of connecting using QRCode as example you will see this can happen.